File: src/ruby/sort_bubble_sort.rb
#####################
# sort script start #
#####################

def do_sorting(input_list)
  output_list = input_list
  output_list = output_list.sort
  output_list
end

###################
# sort script end #
###################

# 👇🏽 copy pasted helpers

input_file_path = ENV["INPUT_PATH"]
input_file = File.readlines(input_file_path)

sorted_data = do_sorting(input_file)

output_file_path = ENV["OUTPUT_PATH"]
File.open(output_file_path, "wb") { |f| f.write(sorted_data.join("")) }

File: src/js/sort_bubble_sort.js
var fs = require("fs");

//////////////////////////
// business logic start //
//////////////////////////

function doSorting(inputList) {
  return bubbleSort(inputList);
}

function bubbleSort(inputList) {
  outputList = inputList;
  isSorted = false;

  while (isSorted === false) {
    outputList, (isSorted = doSortingRound(outputList));
  }

  return outputList;
}

function doSortingRound(inputList) {
  outputList = [];
  isSorted = true;

  inputList.forEach((element, index) => {
    if (index === 0) {
      outputList.push(element);
    } else if (element < outputList[index - 1]) {
      previousElement = outputList[index - 1];
      outputList.pop();
      outputList.push(element);
      outputList.push(previousElement);
      isSorted = false;
    } else {
      outputList.push(element);
    }
  });

  return outputList, isSorted;
}

////////////////////////
// business logic end //
////////////////////////

// 👇🏽 copy pasted helpers

// read input file
let inputFilePath = process.env.INPUT_PATH;
let inputString = fs.readFileSync(inputFilePath, "utf8");
let inputStringSplit = inputString.split(/\n/);

// clean input data
if (inputStringSplit[inputStringSplit.length - 1] === "") {
  inputStringSplit = inputStringSplit.slice(0, inputStringSplit.length - 1);
}

// do sorting
sortedData = doSorting(inputStringSplit);

// write output file
let outputFilePath = process.env.OUTPUT_PATH;
let outputString = sortedData.join("\n");
outputString += "\n"; // trailing newline
fs.writeFileSync(outputFilePath, outputString);
